### PR TITLE
[CodeGen][LLVM] Make the `va_list` related intrinsics generic.

### PR DESCRIPTION
**Pull request by AlexVlx**
Currently, the builtins used for implementing `va_list` handling unconditionally take their arguments as unqualified `ptr`s i.e. pointers to AS 0. This does not work for targets where the default AS is not 0 or AS 0 is not a viable AS (for example, a target might choose 0 to represent the constant address space). This patch changes the builtins' signature to take generic `anyptr` args, which corrects this issue. It is noisy due to the number of tests affected. A test for an upstream target which does not use 0 as its default AS (SPIRV for HIP device compilations) is added as well.

### CODE REVIEW

File: clang/test/CodeGen/varargs-with-nonzero-default-address-space.c
Comment (jrtc27) [inline clang/test/CodeGen/varargs-with-nonzero-default-address-space.c:22]: Needs a newline

File: clang/test/CodeGen/varargs-with-nonzero-default-address-space.c
Comment (jrtc27) [inline clang/test/CodeGen/varargs-with-nonzero-default-address-space.c:1]: Does it actually? Normally -emit-llvm works regardless of the enabled backends

File: clang/test/CodeGen/varargs-with-nonzero-default-address-space.c
Comment (jrtc27) [inline clang/test/CodeGen/varargs-with-nonzero-default-address-space.c:3]: Newline before

File: clang/test/CodeGen/varargs-with-nonzero-default-address-space.c
Comment (jrtc27) [inline clang/test/CodeGen/varargs-with-nonzero-default-address-space.c:19]: Use update_cc_test_checks.py?

File: llvm/include/llvm/IR/Intrinsics.td
Comment (jrtc27) [inline llvm/include/llvm/IR/Intrinsics.td:1719]: Don't touch unrelated whitespace

**Review by jrtc27 — COMMENTED** [review]
Some minor comments, but otherwise would be good to see this upstream. We've had this downstream in CHERI LLVM for many years and never got round to upstreaming it.

**Review by arichardson — COMMENTED** [review]
Thanks, this looks great once all of @jrtc27's comments have been addressed.

**Comment by yxsamliu** [comment]
Is LLVM IR using the old vaarg intrinsics still consumable?

Should we update AutoUpgrade.cpp to auto upgrade the old vaarg intrinsics and add a test for them?

**Comment by jrtc27** [comment]
Doesn’t AutoUpgrade automatically infer overloads? You can see a bunch of tests in this patch where the output references the overloaded intrinsics but the input is unchanged.

**Comment by yxsamliu** [comment]
> Doesn’t AutoUpgrade automatically infer overloads? You can see a bunch of tests in this patch where the output references the overloaded intrinsics but the input is unchanged.

OK it seems to be handled already and test covered. Thanks.

File: clang/test/CodeGen/varargs-with-nonzero-default-address-space.c
Comment (AlexVlx) [reply clang/test/CodeGen/varargs-with-nonzero-default-address-space.c]: Done.

File: clang/test/CodeGen/varargs-with-nonzero-default-address-space.c
Comment (AlexVlx) [reply clang/test/CodeGen/varargs-with-nonzero-default-address-space.c]: I was not certain and erred on the side of caution. You are correct (as far as I can tell), and have removed the requirement. Thank you for pointing it out!

File: llvm/include/llvm/IR/Intrinsics.td
Comment (AlexVlx) [reply llvm/include/llvm/IR/Intrinsics.td]: Done.

File: llvm/include/llvm/IR/Intrinsics.td
Comment (arsenm) [inline llvm/include/llvm/IR/Intrinsics.td:707]: vacopy probably shouldn't have 2 type parameters. 
```suggestion
def int_vacopy  : DefaultAttrsIntrinsic<[],
                                        [llvm_anyptr_ty, LLVMMatchType<0>], [],
                                        "llvm.va_copy">;
```

Comment (AlexVlx) [reply llvm/include/llvm/IR/Intrinsics.td:707]: Hmm, I did struggle a bit with this and admit that I'm not (yet) entirely convinced a valid (albeit hypothetical and admittedly odd) case couldn't be constructed. Consider e.g. having a `__builtin_va_list` declared at namespace / global scope. If a target uses an explicit, non-generic, AS, for globals, then the copy would be to/from a pointer to generic (or pointer to private) from/to a pointer to global. I _believe_ this should work, and making `va_copy` mono-parameter would break it, would it not?

Comment (AlexVlx) [reply llvm/include/llvm/IR/Intrinsics.td:707]: It does for our target, but I believe each target has freedom when it comes to implementing the actual `va_list` underlying type (the standard regards it as opaque), so it is probably legal to do things that do not match that. So I'd rather follow up with some AMDGPU specific work, but leave this as general / generic as possible.

Comment (efriedma-quic) [reply llvm/include/llvm/IR/Intrinsics.td:707]: clang semantic analysis currently forces both the source and destination of va_copy to the type __builtin_va_list_ref, whatever that is.  So clang would never emit va_copy with different pointer types.  (Some other frontend could theoretically emit such operations, but it's hard for me to imagine it being useful.)

In theory, you could try to optimize the address-spaces involved in a va_copy operation... but there's not really much point to trying to optimize va_copy before you lower it.

Comment (AlexVlx) [reply llvm/include/llvm/IR/Intrinsics.td]: Ok, I think you and @arsenm have the right of it, I've taken Matt's suggestion.

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (arichardson) [inline clang/lib/CodeGen/CGBuiltin.cpp:802]:
```suggestion
  // Note: Our __builtin_object_size implementation currently treats Type=0 and
```
typo

**Comment by jrtc27** [comment]
This probably also needs to tweak LangRef

**Review by arsenm — APPROVED** [review]

**Review by arichardson — APPROVED** [review]
LGTM thanks! Please wait for @jrtc27 to also approve before committing though.

File: llvm/test/Bitcode/compatibility-4.0.ll
Comment (jrtc27) [inline llvm/test/Bitcode/compatibility-4.0.ll:1325]: Typo

**Review by jrtc27 — APPROVED** [review]
LGTM once typo is fixed

**Comment by arichardson** [comment]
CI looks unhappy, mlir also seems to need updates:

MLIR :: Target/LLVMIR/llvmir.mlir
MLIR :: mlir-cpu-runner/x86-varargs.mlir

**Comment by AlexVlx** [comment]
> CI looks unhappy, mlir also seems to need updates:
> 
> MLIR :: Target/LLVMIR/llvmir.mlir MLIR :: mlir-cpu-runner/x86-varargs.mlir

Done.

**Review by arichardson — APPROVED** [review]

**Comment by AlexVlx** [comment]
Thank you everyone for the reviews.

**Pull request merged**
